File: remap-customers-after-a-store-move/python/remap_customers.py
"""Remap WooCommerce orders to the correct customer after a store move.

A migration usually re-creates WordPress users with new IDs while orders keep
their old numeric customer_id. This walks every order, finds the WordPress
user whose email matches the order's billing email, cross checks that user's
saved Stripe customer id against the order's Stripe customer id, and only
remaps when both signals agree on exactly one account. Safe to run again and
again; already-correct orders are always skipped. Read only until DRY_RUN is
turned off.

Guide: https://www.allanninal.dev/woocommerce/remap-customers-after-a-store-move/
"""
import os
import logging
import stripe
import requests
from requests.auth import HTTPBasicAuth

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("remap_customers")

stripe.api_key = os.environ.get("STRIPE_SECRET_KEY", "sk_test_dummy")
WOO_URL = os.environ.get("WOO_STORE_URL", "https://example.com").rstrip("/")
AUTH = HTTPBasicAuth(
    os.environ.get("WOO_CONSUMER_KEY", "ck_dummy"),
    os.environ.get("WOO_CONSUMER_SECRET", "cs_dummy"),
)
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"


def all_orders():
    page = 1
    while True:
        r = requests.get(
            f"{WOO_URL}/wp-json/wc/v3/orders",
            params={"per_page": 50, "page": page, "orderby": "id", "order": "asc"},
            auth=AUTH, timeout=30,
        )
        r.raise_for_status()
        batch = r.json()
        if not batch:
            return
        for order in batch:
            yield order
        page += 1


def user_exists(customer_id):
    if not customer_id:
        return False
    r = requests.get(f"{WOO_URL}/wp-json/wc/v3/customers/{customer_id}", auth=AUTH, timeout=30)
    return r.status_code == 200


def users_by_email(email):
    if not email:
        return []
    r = requests.get(
        f"{WOO_URL}/wp-json/wc/v3/customers",
        params={"email": email, "per_page": 10},
        auth=AUTH, timeout=30,
    )
    r.raise_for_status()
    return r.json()


def meta_value(order, key):
    """Read a value out of an order's WooCommerce meta_data list."""
    for meta in order.get("meta_data") or []:
        if meta.get("key") == key and meta.get("value"):
            return meta["value"]
    return None


def stripe_customer_id_of_order(order):
    """The Stripe customer id for an order, from meta or the PaymentIntent."""
    direct = meta_value(order, "_stripe_customer_id")
    if direct:
        return direct
    intent_id = meta_value(order, "_stripe_intent_id") or order.get("transaction_id")
    if intent_id and intent_id.startswith("pi_"):
        try:
            intent = stripe.PaymentIntent.retrieve(intent_id)
            return intent.get("customer")
        except stripe.error.InvalidRequestError:
            return None
    return None


def stripe_ids_agree(order_stripe_customer_id, user_stripe_customer_id):
    """True unless both ids are present and disagree."""
    if not order_stripe_customer_id or not user_stripe_customer_id:
        return True  # nothing to contradict, email match stands
    return order_stripe_customer_id == user_stripe_customer_id


def decide(order, current_customer_valid, matching_users):
    """Pure decision function: no I/O, easy to unit test.

    order: dict with at least "id" and "customer_id".
    current_customer_valid: bool, whether order["customer_id"] resolves to a
        real WooCommerce customer on this site.
    matching_users: list of WooCommerce customer dicts whose email equals the
        order's billing email, each with at least an "id" key.
    """
    if current_customer_valid:
        return ("skip", "customer_id already resolves to a real account")
    if len(matching_users) == 0:
        return ("orphan", "no WordPress account matches this billing email")
    if len(matching_users) > 1:
        return ("ambiguous", "more than one account shares this billing email")
    match = matching_users[0]
    if match["id"] == order.get("customer_id"):
        return ("skip", "already pointing at the matching account")
    return ("remap", f"remap to user {match['id']}")


def remap_order(order, new_customer_id):
    old_id = order.get("customer_id")
    requests.put(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}",
        json={"customer_id": new_customer_id},
        auth=AUTH, timeout=30,
    ).raise_for_status()
    requests.post(
        f"{WOO_URL}/wp-json/wc/v3/orders/{order['id']}/notes",
        json={"note": f"Remapped customer_id from {old_id} to {new_customer_id} after the "
                      f"store move. Matched by Stripe customer id and billing email."},
        auth=AUTH, timeout=30,
    ).raise_for_status()


def run():
    remapped = 0
    reported = 0
    for order in all_orders():
        current_valid = user_exists(order.get("customer_id"))
        email = (order.get("billing") or {}).get("email")
        candidates = users_by_email(email) if not current_valid else []
        action, reason = decide(order, current_valid, candidates)

        if action == "skip":
            continue

        if action in ("orphan", "ambiguous"):
            log.warning("Order %s: %s", order["id"], reason)
            reported += 1
            continue

        match = candidates[0]
        order_stripe_id = stripe_customer_id_of_order(order)
        user_stripe_id = match.get("meta_data_stripe_customer_id") or None
        if not stripe_ids_agree(order_stripe_id, user_stripe_id):
            log.warning("Order %s: Stripe customer id disagrees with email match, skipping", order["id"])
            reported += 1
            continue

        log.info("Order %s: %s. %s", order["id"], reason, "would remap" if DRY_RUN else "remapping")
        if not DRY_RUN:
            remap_order(order, match["id"])
        remapped += 1

    log.info(
        "Done. %d order(s) %s, %d reported for manual review.",
        remapped, "to remap" if DRY_RUN else "remapped", reported,
    )


if __name__ == "__main__":
    run()

File: remap-customers-after-a-store-move/node/remap-customers.js
/**
 * Remap WooCommerce orders to the correct customer after a store move.
 *
 * A migration usually re-creates WordPress users with new IDs while orders
 * keep their old numeric customer_id. This walks every order, finds the
 * WordPress user whose email matches the order's billing email, cross
 * checks that user's saved Stripe customer id against the order's Stripe
 * customer id, and only remaps when both signals agree on exactly one
 * account. Safe to run again and again; already-correct orders are skipped.
 * Read only until DRY_RUN is turned off.
 *
 * Guide: https://www.allanninal.dev/woocommerce/remap-customers-after-a-store-move/
 */
import Stripe from "stripe";
import { pathToFileURL } from "node:url";

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY || "sk_test_dummy");
const WOO_URL = (process.env.WOO_STORE_URL || "https://example.com").replace(/\/$/, "");
const AUTH = "Basic " + Buffer.from(
  `${process.env.WOO_CONSUMER_KEY || "ck_dummy"}:${process.env.WOO_CONSUMER_SECRET || "cs_dummy"}`
).toString("base64");
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

async function woo(path, options = {}) {
  const res = await fetch(`${WOO_URL}/wp-json/wc/v3${path}`, {
    ...options,
    headers: { "Content-Type": "application/json", Authorization: AUTH, ...(options.headers || {}) },
  });
  return res;
}

async function* allOrders() {
  let page = 1;
  while (true) {
    const res = await woo(`/orders?per_page=50&page=${page}&orderby=id&order=asc`);
    if (!res.ok) throw new Error(`Woo /orders returned ${res.status}`);
    const batch = await res.json();
    if (!batch.length) return;
    for (const order of batch) yield order;
    page++;
  }
}

async function userExists(customerId) {
  if (!customerId) return false;
  const res = await woo(`/customers/${customerId}`);
  return res.status === 200;
}

async function usersByEmail(email) {
  if (!email) return [];
  const res = await woo(`/customers?email=${encodeURIComponent(email)}&per_page=10`);
  if (!res.ok) throw new Error(`Woo /customers returned ${res.status}`);
  return res.json();
}

function metaValue(order, key) {
  for (const meta of order.meta_data || []) {
    if (meta.key === key && meta.value) return meta.value;
  }
  return null;
}

async function stripeCustomerIdOfOrder(order) {
  const direct = metaValue(order, "_stripe_customer_id");
  if (direct) return direct;
  const intentId = metaValue(order, "_stripe_intent_id") || order.transaction_id;
  if (intentId && intentId.startsWith("pi_")) {
    try {
      const intent = await stripe.paymentIntents.retrieve(intentId);
      return intent.customer || null;
    } catch {
      return null;
    }
  }
  return null;
}

/** True unless both ids are present and disagree. */
export function stripeIdsAgree(orderStripeCustomerId, userStripeCustomerId) {
  if (!orderStripeCustomerId || !userStripeCustomerId) return true; // nothing to contradict
  return orderStripeCustomerId === userStripeCustomerId;
}

/**
 * Pure decision function: no I/O, easy to unit test.
 *
 * order: object with at least { id, customer_id }.
 * currentCustomerValid: boolean, whether order.customer_id resolves to a
 *   real WooCommerce customer on this site.
 * matchingUsers: array of WooCommerce customer objects whose email equals
 *   the order's billing email, each with at least an `id` field.
 */
export function decide(order, currentCustomerValid, matchingUsers) {
  if (currentCustomerValid) return ["skip", "customer_id already resolves to a real account"];
  if (matchingUsers.length === 0) return ["orphan", "no WordPress account matches this billing email"];
  if (matchingUsers.length > 1) return ["ambiguous", "more than one account shares this billing email"];
  const match = matchingUsers[0];
  if (match.id === order.customer_id) return ["skip", "already pointing at the matching account"];
  return ["remap", `remap to user ${match.id}`];
}

async function remapOrder(order, newCustomerId) {
  const oldId = order.customer_id;
  await woo(`/orders/${order.id}`, {
    method: "PUT",
    body: JSON.stringify({ customer_id: newCustomerId }),
  });
  await woo(`/orders/${order.id}/notes`, {
    method: "POST",
    body: JSON.stringify({
      note: `Remapped customer_id from ${oldId} to ${newCustomerId} after the store move. ` +
            `Matched by Stripe customer id and billing email.`,
    }),
  });
}

export async function run() {
  let remapped = 0;
  let reported = 0;
  for await (const order of allOrders()) {
    const currentValid = await userExists(order.customer_id);
    const email = (order.billing || {}).email;
    const candidates = currentValid ? [] : await usersByEmail(email);
    const [action, reason] = decide(order, currentValid, candidates);

    if (action === "skip") continue;

    if (action === "orphan" || action === "ambiguous") {
      console.warn(`Order ${order.id}: ${reason}`);
      reported++;
      continue;
    }

    const match = candidates[0];
    const orderStripeId = await stripeCustomerIdOfOrder(order);
    const userStripeId = match.meta_data_stripe_customer_id || null;
    if (!stripeIdsAgree(orderStripeId, userStripeId)) {
      console.warn(`Order ${order.id}: Stripe customer id disagrees with email match, skipping`);
      reported++;
      continue;
    }

    console.log(`Order ${order.id}: ${reason}. ${DRY_RUN ? "would remap" : "remapping"}`);
    if (!DRY_RUN) await remapOrder(order, match.id);
    remapped++;
  }
  console.log(`Done. ${remapped} order(s) ${DRY_RUN ? "to remap" : "remapped"}, ${reported} reported for manual review.`);
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}
